// @hubspot/cms-dev-server/dist/island-runtime.js
// ../cms-js-internal/dist/island-runtime/island-runtime.js
var onVisible = (islandEl, callback) => {
  const observer = new IntersectionObserver((entries) => {
    if (entries.some((entry) => entry.isIntersecting)) {
      observer.disconnect();
      callback();
    }
    return;
  });
  for (let i = 0; i < islandEl.children.length; i++) {
    const child = islandEl.children[i];
    observer.observe(child);
  }
};
var MAX_HOSTNAME_LENGTH = 255;
var VALID_HOSTNAME_PATTERN = /^[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/;
var FORBIDDEN_HOSTNAME_CHARS = /[/?#@:&=+$,[\]<>%\s"'`\\]/;
var INVALID_SUBSTRINGS = [
  "://",
  "?",
  "#",
  ";",
  "\\",
  "..",
  "//",
  // eslint-disable-next-line no-script-url
  "javascript:",
  "data:",
  "vbscript:",
  "<script",
  "base64",
  "alert(",
  "eval(",
  "Function("
];
function isValidHostname(hostname) {
  const isLengthOutOfRange = hostname.length < 1 || hostname.length > MAX_HOSTNAME_LENGTH;
  if (!hostname || typeof hostname !== "string" || isLengthOutOfRange || !hostname.includes(".") || FORBIDDEN_HOSTNAME_CHARS.test(hostname) || !VALID_HOSTNAME_PATTERN.test(hostname) || INVALID_SUBSTRINGS.some(
    (sub) => hostname.toLowerCase().includes(sub.toLowerCase())
  ) || /^[0-9.]+$/.test(hostname)) {
    return false;
  }
  return hostname.split(".").every(
    (label) => label.length >= 1 && label.length <= 63 && !label.startsWith("-") && !label.endsWith("-")
  );
}
function getSecureOverrideUrl(path) {
  const currentUrl = new URL(window.location.toString());
  const overrideHost = currentUrl.searchParams.get("hsDebugOverridePublicHost");
  try {
    if (!overrideHost) {
      throw new Error("No override host provided");
    }
    if (!isValidHostname(overrideHost)) {
      throw new Error("Invalid hostname format");
    }
    const overriddenUrl = new URL(
      `${currentUrl.protocol}//${overrideHost}${path}`
    );
    if (!["http:", "https:"].includes(overriddenUrl.protocol)) {
      throw new Error("Invalid protocol");
    }
    return overriddenUrl.toString();
  } catch (e) {
    console.warn(`Invalid debug override host: ${overrideHost}`, e);
    return `${window.location.origin}${path}`;
  }
}
var rootsToProcess;
var hydrationQueue = [];
var hydrationTypesToAccept = /* @__PURE__ */ new Set(["load", "visible"]);
var startHydratingIdleRootsTimeout;
var shouldForceSeparateJsForAllIslands = /(&|\?)forceSeparateIslandJs=true/.test(window.location.search);
var scheduledHydrationIndexCtr = 1;
var rootsById = /* @__PURE__ */ new Map();
var scheduledElements = /* @__PURE__ */ new WeakSet();
var typedWindow = window;
function findIslandElements(islandId) {
  if (typeof CSS !== "undefined" && CSS.escape) {
    return Array.from(
      document.querySelectorAll(
        `[data-hs-island][id="${CSS.escape(islandId)}"]`
      )
    );
  }
  return Array.from(
    document.querySelectorAll("[data-hs-island]")
  ).filter((el) => el.id === islandId);
}
function prepareAndScheduleRoot(root) {
  root.hydrationStatus = "HYDRATION_PREPARING";
  waitThenLoadAndHydrateIfNeeded(
    root,
    getLoadableModuleFor(root),
    root.wrapperUrl ? getLoadableModuleFor({
      url: root.wrapperUrl,
      id: `${root.id}-wrapper`,
      moduleName: `${root.moduleName}-wrapper`
    }) : void 0
  );
}
function startHydratingIdleRoots() {
  hydrationTypesToAccept.add("idle");
  for (const root of rootsToProcess) {
    if (root.hydrationStatus === "HYDRATION_WAITING" && hydrationTypesToAccept.has(root.hydrateOn)) {
      prepareAndScheduleRoot(root);
    }
  }
}
function setupIdleCallbackIfNeeded() {
  if (!startHydratingIdleRootsTimeout) {
    if (!window.requestIdleCallback) {
      setTimeout(() => {
        startHydratingIdleRoots();
      }, 1);
      return;
    }
    startHydratingIdleRootsTimeout = requestIdleCallback(
      startHydratingIdleRoots,
      { timeout: 5e3 }
    );
  }
}
function finishHydratingRootAndScheduleNext(entry) {
  entry.root.hydrationStatus = "HYDRATED";
  hydrationQueue = hydrationQueue.filter((x) => x !== entry);
  const hasOtherRootsToHydrate = hydrationQueue.length > 0;
  if (hasOtherRootsToHydrate) {
    setTimeout(() => hydrateRoot(hydrationQueue[0]), 0);
  } else if (!hasOtherRootsToHydrate && !hydrationTypesToAccept.has("idle")) {
    setupIdleCallbackIfNeeded();
  }
}
function hydrateRoot(entry) {
  const { rootEl, root, IslandModule, WrapperModule } = entry;
  if (window.location.search.includes("hsDebug=")) {
    console.log("Hydrating root", root.id, root.hydrateOn);
  }
  const postHydration = () => {
    finishHydratingRootAndScheduleNext(entry);
  };
  const renderableIsland = {
    rootElement: rootEl,
    pageUrl: typedWindow.__hsServerPageUrl,
    basePath: typedWindow.__hsBasePath,
    islands: typedWindow.__islands,
    postHydration,
    island: root,
    WrapperModule: WrapperModule ? WrapperModule.default : void 0,
    index: entry.islandScheduledIndex
  };
  const reactRoot = IslandModule.render(renderableIsland);
  const existingRoots = rootsById.get(root.id) || [];
  rootsById.set(root.id, [...existingRoots, reactRoot]);
}
function pushToHydrationQueue(entry) {
  entry.root.hydrationStatus = "HYDRATION_QUEUED";
  hydrationQueue.push(entry);
  if (hydrationQueue.length === 1) {
    hydrateRoot(entry);
  }
}
function loadThenHydrate(root, LoadableIslandModule, LoadableWrapperModule, rootEl, islandScheduledIndex) {
  Promise.all([
    LoadableIslandModule(),
    LoadableWrapperModule ? LoadableWrapperModule() : Promise.resolve()
  ]).then(
    ([IslandModule, WrapperModule]) => {
      pushToHydrationQueue({
        rootEl,
        root,
        IslandModule,
        WrapperModule,
        islandScheduledIndex
      });
    }
  ).catch((error) => {
    console.error(`Failed to load module for island: ${root.id}`);
    console.error(error);
    scheduledElements.delete(rootEl);
  });
}
function waitThenLoadAndHydrateIfNeeded(root, LoadableIslandModule, LoadableWrapperModule) {
  const rootEls = findIslandElements(root.id);
  rootEls.forEach((rootEl) => {
    if (scheduledElements.has(rootEl)) {
      return;
    }
    scheduledElements.add(rootEl);
    const rootClone = { ...root };
    const islandScheduledIndex = scheduledHydrationIndexCtr;
    scheduledHydrationIndexCtr += 1;
    if (rootClone.hydrateOn === "visible") {
      onVisible(rootEl, () => {
        loadThenHydrate(
          rootClone,
          LoadableIslandModule,
          LoadableWrapperModule,
          rootEl,
          islandScheduledIndex
        );
      });
    } else {
      loadThenHydrate(
        rootClone,
        LoadableIslandModule,
        LoadableWrapperModule,
        rootEl,
        islandScheduledIndex
      );
    }
  });
}
function getLoadableModuleFor(root) {
  let url = root.url;
  if (root.url.startsWith("/") && !root.url.startsWith("//") && window.location.search.includes("hsDebugOverridePublicHost=")) {
    url = getSecureOverrideUrl(root.url);
  } else if (!root.url.startsWith("https://") && !root.url.startsWith("http://") && !root.url.startsWith("//")) {
    url = `${window.location.origin}${root.url}`;
  }
  if (shouldForceSeparateJsForAllIslands) {
    url = `${root.url}?cacheBust=${Date.now()}-${root.id}`;
  }
  return () => import(
    /* @vite-ignore */
    url
  );
}
function hydrateIslandRoots() {
  if (rootsToProcess === void 0) {
    rootsToProcess = typedWindow.__islands;
    if (window.location.search.includes("hsDebug=")) {
      console.log("Starting to hydrate", rootsToProcess);
    }
    if (shouldForceSeparateJsForAllIslands) {
      console.log("  ... forcibly making every island refetch JS");
    }
  }
  let encounteredAnyNonIdleIslands = false;
  for (const root of rootsToProcess) {
    if (!root.hydrationStatus) {
      if (hydrationTypesToAccept.has(root.hydrateOn)) {
        encounteredAnyNonIdleIslands = true;
        prepareAndScheduleRoot(root);
      } else {
        root.hydrationStatus = "HYDRATION_WAITING";
      }
    }
  }
  if (rootsToProcess.length > 0 && !encounteredAnyNonIdleIslands) {
    setupIdleCallbackIfNeeded();
  }
}
function setupIslandHydration() {
  if (document.readyState === "interactive" || document.readyState === "complete") {
    hydrateIslandRoots();
  } else {
    window.addEventListener("DOMContentLoaded", hydrateIslandRoots);
  }
}
function unmountIsland(islandId) {
  const islandToUnmount = rootsToProcess.find(
    (island) => island.id === islandId
  );
  const reactRoots = rootsById.get(islandId);
  if (!islandToUnmount || !reactRoots || reactRoots.length === 0) {
    throw new Error(`No known island: ${islandId}`);
  }
  const islandIndex = rootsToProcess.indexOf(islandToUnmount);
  rootsToProcess.splice(islandIndex, 1);
  rootsById.delete(islandId);
  const islandEls = findIslandElements(islandId);
  islandEls.forEach((el) => scheduledElements.delete(el));
  console.log("Unmounting island:", islandId);
  for (const reactRoot of reactRoots) {
    reactRoot.unmount();
  }
}
export {
  setupIslandHydration,
  unmountIsland
};
